packages/runtime/container-runtime/src/containerRuntime.ts

```diff
@@ -2789,118 +2777,61 @@ export class ContainerRuntime
-		this.scheduleManager.beforeOpProcessing(keyMessage);
-
-		this._processedClientSequenceNumber = batchStartCsn;
-		if (!this.hasPendingMessages()) {
-			this.updateDocumentDirtyState(false);
-		}
-
-		// We emit this event but say isRuntimeMessage is false, because there are no actual runtime messages here being processed.
-		// But someone listening to this event expecting to be notified whenever a message arrives would want to know about this.
-		this.emit("op", keyMessage, false /* isRuntimeMessage */);
-		this.scheduleManager.afterOpProcessing(undefined /* error */, keyMessage);
-	}
-
-	/**
-	 * Observes messages that are not intended for the runtime layer, updating/notifying Runtime systems as needed.
-	 * @param messageWithContext - non-runtime messages to process with additional context and isRuntimeMessage prop as false
-	 */
-	private observeNonRuntimeMessage(
-		messageWithContext: MessageWithContext & { isRuntimeMessage: false },
-	) {
-		const { message } = messageWithContext;
-
-		// Intercept to reduce minimum sequence number to the delta manager's minimum sequence number.
-		// Sequence numbers are not guaranteed to follow any sort of order. Re-entrancy is one of those situations
-		if (
-			this.deltaManager.minimumSequenceNumber <
-			messageWithContext.message.minimumSequenceNumber
-		) {
-			messageWithContext.message.minimumSequenceNumber =
-				this.deltaManager.minimumSequenceNumber;
-		}
-
-		// Surround the actual processing of the operation with messages to the schedule manager indicating
-		// the beginning and end. This allows it to emit appropriate events and/or pause the processing of new
-		// messages once a batch has been fully processed.
-		this.scheduleManager.beforeOpProcessing(message);
+			batch.forEach(({ message, localOpMetadata }) => {
+				this.ensureNoDataModelChanges(() => {
+					// Set the minimum sequence number to the containerRuntime's understanding of minimum sequence number.
+					if (this.deltaManager.minimumSequenceNumber < message.minimumSequenceNumber) {
+						message.minimumSequenceNumber = this.deltaManager.minimumSequenceNumber;
+					}
 
-		this._processedClientSequenceNumber = message.clientSequenceNumber;
+					this._processedClientSequenceNumber = message.clientSequenceNumber;
 
-		try {
-			// If there are no more pending messages after processing a local message,
-			// the document is no longer dirty.
-			if (!this.hasPendingMessages()) {
-				this.updateDocumentDirtyState(false);
-			}
+					// If there are no more pending messages after processing a local message,
+					// the document is no longer dirty.
+					if (!this.hasPendingMessages()) {
+						this.updateDocumentDirtyState(false);
+					}
 
-			this.emit("op", message, messageWithContext.isRuntimeMessage);
+					if (runtimeBatch) {
+						this.validateAndProcessRuntimeMessage({
+							message: message as InboundSequencedContainerRuntimeMessage,
+							local,
+							savedOp,
+							localOpMetadata,
+						});
+					}
```

Comment:
I think I can make the case that none of this code needs to run for non-runtime messages... I like the refactor to combine the flows to this one choke point, and I suppose if we want to remove this unneeded code execution we can pull it into the `if` block.

Reply:
I am inclined to agree. The only parts worth keeping for non-runtime messages are emitting "batchBegin", "batchEnd" and "op" events.
However, for this PR, I brought back the separate processing of runtime vs non-runtime messages. Maybe I will do a follow up to clean it up.